asynction/security.py

```diff
@@ -0,0 +1,401 @@
+    if scheme.scheme == HTTPAuthenticationScheme.BASIC:
+        basic_info_func = load_basic_info_func(scheme)
+        scope_validate_func = load_scope_validate_func(scheme)
+
+        def http_security_check(request: Request) -> InternalSecurityCheckResponse:
+            token_info = validate_basic(request, basic_info_func, required_scopes)
+            if token_info is None:
+                return None, None
+
+            token_scopes = token_info.get("scope", token_info.get("scopes", ""))
+
+            if not scope_validate_func(required_scopes, token_scopes):
+                return (
+                    None,
+                    f"Invalid scopes: required: {required_scopes}, provided: {token_scopes}",  # noqa: 501
+                )
+
+            return token_info, None
+
+        return http_security_check
+    elif scheme.scheme == HTTPAuthenticationScheme.BEARER:
+        api_key_info_func = load_api_key_info_func(scheme)
+        scope_validate_func = load_scope_validate_func(scheme)
+
+        bearer_format = scheme.bearer_format
+
+        def http_bearer_security_check(
+            request: Request,
+        ) -> InternalSecurityCheckResponse:
+            token_info = validate_api_key(
+                request, api_key_info_func, required_scopes, bearer_format
+            )
+            if token_info is None:
+                return None, None
+
+            token_scopes = token_info.get("scope", token_info.get("scopes", ""))
+
+            if not scope_validate_func(required_scopes, token_scopes):
+                return (
+                    None,
+                    f"Invalid scopes: required: {required_scopes}, provided: {token_scopes}",  # noqa: 501
+                )
+
+            return token_info, None
+
+        return http_bearer_security_check
+    else:
+        return None
+
+
+def get_cookie_value(cookies: str, name: str) -> Optional[str]:
+    """
+    Returns cookie value by its name. None if no such value.
+    :param cookies: str: cookies raw data
+    :param name: str: cookies key
+
+    Borrowed from https://github.com/zalando/connexion/blob/main/connexion/security/security_handler_factory.py#L206  # noqa: 501
+    """
+    cookie_parser: SimpleCookie = SimpleCookie()
+    cookie_parser.load(str(cookies))
+    try:
+        return cookie_parser[name].value
+    except KeyError:
+        return None
+
+
+def build_http_api_key_security_check(
+    requirement: InternalSecurityRequirement, scheme: SecurityScheme
+) -> Optional[InternalSecurityCheck]:
+    api_key_info_func = load_api_key_info_func(scheme)
+    scope_validate_func = load_scope_validate_func(scheme)
+
+    _, required_scopes = requirement
+    api_key_in = scheme.in_
+    api_key_name = scheme.name
+    if api_key_name is None:
+        raise SecurityException("invalid api key name specified")
+    if api_key_in not in ["query", "header", "cookie"]:
+        raise SecurityException("invalid api key location specified")
```

Comment:
These 2 validations should be moved in the `types.py` module, under the `__post_init__` method of the `SecurityScheme` dataclass.

Reply:
Also, within `types.py`, let's deserialise `query`, `header`, `cookie` as an enum.

For example:
```python
class ApiKeyLocation(Enum):
    USER = "user"
    PASSWORD = "password"
    QUERY = "query"
    HEADER = "header"
    COOKIE = "cookie"
```